ledger/forest/forest.go

```diff
@@ -0,0 +1,194 @@
+package forest
+
+import (
+	"fmt"
+
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/ledger/common/hash"
+	"github.com/onflow/flow-go/model/flow"
+
+	"github.com/optakt/flow-dps/ledger/trie"
+)
+
+type step struct {
+	tree   *trie.Trie
+	parent flow.StateCommitment
+}
+
+// TODO: Look into replacing the forest with a Forest where possible, and improve the Forest
+//  to be more performant overall. It could also make sense however to keep this forest implementation
+//  for our case and keep another one with more features around.
+
+// Forest is a forest of tries with an unlimited size, but which can be reset manually to remove all tries besides one.
+type Forest struct {
+	values map[ledger.Path]*ledger.Payload
+	steps  map[flow.StateCommitment]step
+}
+
+func New() *Forest {
+	f := Forest{
+		steps:  make(map[flow.StateCommitment]step),
+		values: make(map[ledger.Path]*ledger.Payload),
+	}
+
+	return &f
+}
+
+func (f *Forest) Add(tree *trie.Trie, paths []ledger.Path, payloads []*ledger.Payload, parent flow.StateCommitment) {
+	commit := flow.StateCommitment(tree.RootHash())
+
+	// FIXME: Is this still necessary? IIRC it was a temporary solution because we didn't have access to payloads anymore. Since we can list the paths and have access to the DB, we might do that instead of storing values in there.
+	for i := range paths {
+		if payloads == nil {
+			f.values[paths[i]] = nil
+		} else {
+			f.values[paths[i]] = payloads[i]
+		}
+	}
+
+	s := step{
+		tree:   tree,
+		parent: parent,
+	}
+	f.steps[commit] = s
+}
+
+func (f *Forest) Has(commit flow.StateCommitment) bool {
+	_, ok := f.steps[commit]
+	return ok
+}
+
+func (f *Forest) Tree(commit flow.StateCommitment) (*trie.Trie, bool) {
+	s, ok := f.steps[commit]
+	if !ok {
+		return nil, false
+	}
+
+	return s.tree, true
+}
+
+func (f *Forest) Parent(commit flow.StateCommitment) (flow.StateCommitment, bool) {
+	st, ok := f.steps[commit]
+	if !ok {
+		return flow.DummyStateCommitment, false
+	}
+
+	return st.parent, true
+}
+
+func (f *Forest) Values() map[ledger.Path]*ledger.Payload {
+	return f.values
+}
+
+func (f *Forest) Reset(finalized flow.StateCommitment) {
+	for commit := range f.steps {
+		if commit != finalized {
+			delete(f.steps, commit)
+		}
+	}
+}
+
+// FIXME: The following is only needed for this struct to be used as a flow forest.
+
+func (f *Forest) Read(r *ledger.TrieRead) ([]*ledger.Payload, error) {
+
+	if len(r.Paths) == 0 {
+		return []*ledger.Payload{}, nil
+	}
+
+	commit, err := flow.ToStateCommitment(r.RootHash[:])
+	if err != nil {
+		return nil, fmt.Errorf("invalid trie root hash: %w", err)
+	}
+
+	trie, ok := f.Tree(commit)
+	if !ok {
+		return nil, fmt.Errorf("trie does not exist (root hash: %x)", commit[:])
+	}
+
+	// Deduplicate keys.
+	deduplicatedPaths := make([]ledger.Path, 0, len(r.Paths))
```

Comment:
This is something that Flow consistently does whenever they deal with a batch of updates, so I assume there is a legitimate concern for it. I am not really familiar with what can and cannot happen in _the real world_ but I'll have a meeting with them this evening and I can raise this point. I assume that it's there more to prevent edge cases than to improve performance though, TBH. Since they concurrently insert values in the trie, I can imagine that having two of the same path could cause panics.